client/asset/eth/contractor.go

```diff
@@ -262,6 +277,151 @@ func (c *contractorV0) incomingValue(ctx context.Context, tx *types.Transaction)
 	return swap.Value, nil
 }
 
+type contractorV1 struct {
+	contractV1
+	abi          *abi.ABI
+	net          dex.Network
+	contractAddr common.Address
+	acctAddr     common.Address
+	be           bind.ContractBackend
+}
+
+// var _ contractor = (*contractorV1)(nil)
+
+func newV1contractor(net dex.Network, acctAddr common.Address, ec *ethclient.Client) (*contractorV1, error) {
+	contractAddr, exists := dexeth.ContractAddresses[1][net]
+	if !exists || contractAddr == (common.Address{}) {
+		return nil, fmt.Errorf("no contract address for version 0, net %s", net)
+	}
+	c, err := swapv1.NewETHSwap(contractAddr, ec)
+	if err != nil {
+		return nil, err
+	}
+	return &contractorV1{
+		contractV1:   c,
+		abi:          dexeth.ABIs[1],
+		net:          net,
+		contractAddr: contractAddr,
+		acctAddr:     acctAddr,
+		be:           ec,
+	}, nil
+}
+
+func (c *contractorV1) swap(ctx context.Context, secretHash [32]byte) (*dexeth.SwapState, error) {
+	return nil, nil
+}
+
+func (c *contractorV1) initiate(txOpts *bind.TransactOpts, contracts []*asset.Contract) (*types.Transaction, error) {
+	versionedContracts := make([]swapv1.ETHSwapContract, 0, len(contracts))
+	for _, contract := range contracts {
+		versionedContracts = append(versionedContracts, ContractToV1(c.acctAddr, contract))
+	}
+	return c.Initiate(txOpts, versionedContracts)
+}
+
+func (c *contractorV1) redeem(txOpts *bind.TransactOpts, redeems []*asset.Redemption) (*types.Transaction, error) {
+	versionedRedemptions := make([]swapv1.ETHSwapRedemption, 0, len(redeems))
+	for _, r := range redeems {
+		var initiator common.Address
+		copy(initiator[:], r.Spends.Coin.ID())
+		versionedRedemptions = append(versionedRedemptions, RedemptionToV1(initiator, c.acctAddr, r))
+	}
+	return c.Redeem(txOpts, versionedRedemptions)
+}
+
+// This one breaks contractor, since we need the entire *asset.Contract, not just
+// the secret hash.
```

Comment:
There are quite a few broken like this, including `swap`, which is used all over. e.g. https://github.com/decred/dcrdex/pull/1638/files#diff-a45b0f60d5cf2f6825eb9e9ed1f173fbf1df9eb53b09914ce6b17e42074653e5R101

I'm confused about how we're going to adapt if we can't look up the `Swap struct` by secret hash any more.

Reply:
> I'm confused about how we're going to adapt if we can't look up the Swap struct by secret hash any more.

We could do it with the `Swaps` and the `ContractKey` methods. We would just need to interpret ourselves. But I've added a little more detail to the return from state (renamed to status) to make things easier. Happy to go back if you're concerned with the contract change.

Reply:
> It also becomes super hard to use the refund-by-anyone path we are relying on as a backup. When we create the swap we can't a pre-signed refund, so we have been logging the secret hash so the user can just (e.g.) take that to etherscan and refund with metamask.

I'll look into a solution, but I don't see any reason that something like this should stop us.